---
title: "TreeSe-basics"
author: "Jayaram Kancherla, Hector Corrada Bravo"
date: "November 13, 2018"
output: BiocStyle::html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## loading required packages

```{r load-packages, message=FALSE, warning=FALSE}
library(TreeIndex)
library(metagenomeSeq)
library(msd16s)
library(S4Vectors)
```

## Example Data

```{r}
data(mouseData)
counts <- MRcounts(mouseData)
hierarchy <- fData(mouseData)
```


## creating a `TreeSE` object

```{r} 
tree <- TreeIndex(hierarchy)

mbiome <- TreeSE(SimpleList(counts=counts), rowData=tree)
mbiome
```


## Operations on TreeSE objects

`TreeSE` provides `aggregateTree` method, to aggregate counts (rows or columns) to a given tree selection. The result is a `TreeSE` object. In a similar way, one can perform these operations on single cell datasets where colData has a hierarchy.

```{r}
aggr <- aggregateTree(mbiome, selectedLevel=3, by="row")
aggr
```

### Defining a cut in the tree

In the previous example, we use the selectedLevel parameter to cut the tree at a given level 3.  Each node can be in one of three possible states removed (0), expanded (1) and aggregated (2). using this, we can cut a tree at different levels.

Note: A `facetZoom` visualization from `metavizr` will help make node selections easier

```{r}

# lets get the list of nodes

nodes <- getNodes(tree)

# select/remove rows in data frame to only keep nodes to aggregate 
# this example chooses all level 3 nodes and 2 nodes from level 4 (Bacillales and Lactobacillales) 
# to define the cut at different levels
nodes <- nodes[level %in% c(3,4),][1:19]


# aggregate counts
agg_sel <- aggregateTree(mbiome, selectedLevel=3, selectedNodes=nodes, by="row")
agg_sel
```

