---
title: "TreeSe-basics"
author: "Jayaram Kancherla, Hector Corrada Bravo"
date: "November 13, 2018"
output: BiocStyle::html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## loading required packages

```{r load-packages, message=FALSE, warning=FALSE}
library(TreeIndex)
library(metagenomeSeq)
library(msd16s)
library(S4Vectors)
library(clustree)
library(Seurat)
```

## Example Data

```{r}
data(mouseData)
counts <- MRcounts(mouseData)
hierarchy <- fData(mouseData)
```


## creating a `TreeSE` object

```{r} 
tree <- TreeIndex(hierarchy)

mbiome <- TreeSE(SimpleList(counts=counts), rowData=tree)
mbiome
```


## Operations on TreeSE objects

`TreeSE` provides `aggregateTree` method, to aggregate counts (rows or columns) to a given tree selection. The result is a `TreeSE` object. In a similar way, one can perform these operations on single cell datasets where colData has a hierarchy.

```{r}
aggr <- aggregateTree(mbiome, selectedLevel=3, by="row")
aggr
```

### Defining a cut in the tree

In the previous example, we use the selectedLevel parameter to cut the tree at a given level 3.  Each node can be in one of three possible states removed (0), expanded (1) and aggregated (2). using this, we can cut a tree at different levels.

Note: A `facetZoom` visualization from `metavizr` will help make node selections easier

```{r}

# lets get the list of nodes

nodes <- getNodes(tree)

# select/remove rows in data frame to only keep nodes to aggregate 
# this example chooses all level 3 nodes and 2 nodes from level 4 (Bacillales and Lactobacillales) 
# to define the cut at different levels
nodes <- nodes[level %in% c(3,4),][1:19]


# aggregate counts
agg_sel <- aggregateTree(mbiome, selectedLevel=3, selectedNodes=nodes, by="row")
agg_sel
```


### Import a `Seurat` and `clustree` into `TreeSE`

The previous example use a metagenomic dataset. In this section, we will show how to import a Seurat object along with the cluster hierarchy generated from the clustree into a TreeSE object

```{r}
# Seurat tutorial on pbmc 3k dataset
# data set link - https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz
pbmc.data <- Read10X(data.dir = "data/filtered_gene_bc_matrices/hg19/")

pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 3, min.genes = 200,
                           project = "10X_PBMC")

mito.genes <- grep(pattern = "^MT-", x = rownames(x = pbmc@data), value = TRUE)
percent.mito <- Matrix::colSums(pbmc@raw.data[mito.genes, ])/Matrix::colSums(pbmc@raw.data)

pbmc <- AddMetaData(object = pbmc, metadata = percent.mito, col.name = "percent.mito")

pbmc <- FilterCells(object = pbmc, subset.names = c("nGene", "percent.mito"),
                    low.thresholds = c(200, -Inf), high.thresholds = c(2500, 0.05))

pbmc <- NormalizeData(object = pbmc, normalization.method = "LogNormalize",
                      scale.factor = 10000)

pbmc <- FindVariableGenes(object = pbmc, mean.function = ExpMean, dispersion.function = LogVMR,
                          x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)

pbmc <- ScaleData(object = pbmc, vars.to.regress = c("nUMI", "percent.mito"))

pbmc <- RunPCA(object = pbmc, pc.genes = pbmc@var.genes, do.print = TRUE, pcs.print = 1:5,
               genes.print = 5)

pbmc <- FindClusters(object = pbmc, reduction.type = "pca", dims.use = 1:10,
                     resolution = c(0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0), print.output = 0, save.SNN = TRUE)

graph <- clustree(pbmc, return="graph")

pbmc_TreeSE <- ImportFromSeurat(pbmc, graph)
pbmc_TreeSE

```

We can now perform similar Tree operations on the colData similar to the rowData from the previous examples. 
